### DISCUSSION TITLE
Tanstack Start: using <style> tag in <Head> in `__root.tsx` file causes duplication lead to `Uncaught Error: Hydration failed because ...`

### DISCUSSION THREAD

**Discussion by Foxeye-Rinx** [TanStack/router#2587] [Q&A]
<img width="586" alt="image" src="https://github.com/user-attachments/assets/46763647-b6cb-43f1-873d-6938ec5f5615">

The duplication leads to: 
```
Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```

**Context:** 
I just finished the walkthrough of Tanstack Start and trying to add a 404 page with some tailwind styles.
I noticed that the HTML is the first thing that is transferred to the browser. Then `client.tsx` and `app.css` come after.
It causes a noticeable FOUC issue because the CSS comes later.

**The common solution**:
I searched the internet and the common solution is just to hide the`html` with CSS in `<head>` like 
```html
<style id='hideHTML'>
 html { visibility: hidden; opacity: 0; }
</style>
```
And then remove it after the app.css is loaded.

My RootDocument 

```jsx
function RootDocument({ children }: { children: React.ReactNode }) {
  useEffect(() => {
    debugger
    console.log("See duplicated style tags with the debugger above")
  }, [])
  return (
    <Html>
      <Head>
        <style id='hideHTML'>{`html{visibility:hidden;opacity:0;}`}</style>
        <Meta />
      </Head>
      <Body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </Body>
    </Html>
  )
}
```

The `app.css` file load just after the NoutFound component's html  (`/1243421` returns not found component )
<img width="575" alt="image" src="https://github.com/user-attachments/assets/7ba1fdf9-0528-4fe8-83a7-2d967ba84946">

My `app.css`
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

html {
  visibility: visible !important;
  opacity: 1 !important;
}

```
The whole code:
https://github.com/Foxeye-Rinx/TanstackStartWithDuplicatedTags

**More information:**
- Remix tried to make a fix, but it introduced another problem with CSS ordering, which might be a potential bug in building production code:
https://youtu.be/mWK3Y_1kmaM?t=30628
- Astro `is:inline` directive is an approach that allows me more control over how I can handle this FOUC issue
https://docs.astro.build/en/reference/directives-reference/#isinline, Astro removes the <style> tags and bundle it to external files, using `is:inline` is to keep the <style> tag stick to the original HTML, it's a flag to tell the bundler that should not touch to this tag.

**Comment by SoloReverse** [comment]
So I am suffering with duplication using a <meta/> tag that is being rendered twice. I think the issue is with the ```<Meta/>``` component. Its causing all these hydration issues as well.

**Reply by SoloReverse** [reply]
Okay I found the following. The hydration issue was my fault, caused by an image component. 

Moreover the duplicate tags for me were caused by React 19. I think it's too new. I'll probably switch to some other react-helmet and continue using react 18.3.1 until there is a fix.
